[e2e] Update create simple product test

alvarothomas opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

Add the following steps to the test that checks the creation of a simple product:
- add a description
- add a sale price
- add SKU
- include a purchase
- add an attribute (Color)

Closes [woocommerce/woocommerce-quality/issues/607](https://github.com/woocommerce/woocommerce-quality/issues/607)

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Please run the _create simple product test_ as follows:

`pnpm test:e2e-pw ./tests/e2e-pw/tests/merchant/create-simple-product.spec.js`

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [x] Automatically create a changelog entry from the details below.

<details>

#### Significance

<!-- Choose only one -->

-   [x] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [x] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

Updates the test of a virtual product creation functionality with further characteristics

#### Comment <!-- If the changes in this pull request don't warrant a changelog entry, you can alternatively supply a comment here. Note that comments are only accepted with a significance of "Patch" -->

</details>

Commit:
Update create simple product

Add steps that add description, sale price, SKU, purchase note and an attribute (Color)

veljkho commented on plugins/woocommerce/tests/e2e-pw/tests/merchant/create-simple-product.spec.js:
```diff
@@ -4,6 +4,12 @@ const wcApi = require( '@woocommerce/woocommerce-rest-api' ).default;
 const productPrice = '9.99';
+const salePrice = '6.99';
+const productDescription = 'Virtual product description.';
+const productSKU = '1234567890'
+const productPurchaseNote = 'Virtual product purchase note'
+const productAttribute = 'color'
+const productAttributeColor = 'red | white'
```
There are no ending `;` for these lines.
I have prettier installed in my IDE which automatically solves code formatting on save. I also see other formatting issues down below but I will comment it separately.

rodelgc replied:
As a bit of a context, we use the [@wordpress/prettier-config](https://www.npmjs.com/package/@wordpress/prettier-config) package for formatting. When setting up the local testing environment, this package will be installed as part of the `pnpm install` command in the set up steps. So you'll just have to press `Shift + Option + F ( ⇧⌥F )` and WP's own prettier config will fix most of the formatting issues Veljko raised.

alvarothomas replied:
Thank you for pointing this out @veljkho , after +2 years using Python only I need to get used to JS formatting.

alvarothomas replied:
Also, I'll try with the Prettier extension too. Thanks!

veljkho commented on plugins/woocommerce/tests/e2e-pw/tests/merchant/create-simple-product.spec.js:
```diff
@@ -57,7 +63,33 @@ test.describe.serial( 'Add New Simple Product Page', () => {
+			.getByRole( 'link', { name: 'Attributes' } ).click();
+		await page.getByPlaceholder ( 'f.e. size or color' ).fill( productAttribute );
+		await page.getByPlaceholder ( 'Enter some descriptive text. Use “|” to separate different values.' ).fill( productAttributeColor );
+		await page.keyboard.press('Enter');
```
87 and 88 are very long. 89 needs spacing around `Enter`. These are solvable by a prettier formatted.
Please let me know if you need any help around it.

alvarothomas replied:
Tried Prettier to format those. Nice one!

Stojdza commented on plugins/woocommerce/tests/e2e-pw/tests/merchant/create-simple-product.spec.js:
```diff
@@ -57,7 +63,33 @@ test.describe.serial( 'Add New Simple Product Page', () => {
+
+		// Fill in SKU
+		await page.getByText( 'Inventory' ).click();
+		await page.locator( '#_sku' ).fill( productSKU );
```
I've followed @lanej0's [suggestion](https://github.com/woocommerce/woocommerce/pull/43710#discussion_r1454084102) and used `getBy*` instead of `locator` wherever possible. For the SKU form, you can try the same I did, per @veljkho's advice, [here](https://github.com/woocommerce/woocommerce/pull/43710/files#diff-4d8b4fa2d3fc09e284b17868562b01cf7bc2d5294477590b31e45f0f01277ebbR156), it should work 🤞

lanej0 replied:
This looks good @Stojdza.

The failing CI check you have is just because you have to add a changelog entry.  If you go to the `plugins/woocommerce` folder and run `pnpm --filter=plugin-woocommerce run changelog add` you can do that, commit it and it should resolved the failing check.

alvarothomas replied:
Thanks @lanej0, I've just added this via CLI.

Stojdza reviewed (commented):
Left one comment for the SKU field

veljkho commented on plugins/woocommerce/tests/e2e-pw/tests/merchant/create-simple-product.spec.js:
```diff
@@ -57,7 +63,33 @@ test.describe.serial( 'Add New Simple Product Page', () => {
 		await page.locator( '#_virtual' ).click();
+
+		// Fill in a product description
+		await page
+			.frameLocator( '#content_ifr' )
+			.locator( '.wp-editor' )
+			.fill( productDescription );
```
It didn't work for me locally and I realised why.. I had Text view open by default and I don't know why. I'm afraid someone when testing the suite locally could have the same issue or this may produce false-positive.

Wouldn't you mind adding clicking Visual first before filling product description?

```
await page.getByRole( 'button', { name: 'Visual', exact: true } ).first().click();
```

rodelgc replied:
The locator also didn't work for me when the "Product description" field was set to "Text". Would be great to add a step to click on the "Visual" tab first to ensure robustness of the product description locator, as Veljko mentioned above. 

Alternatively, you may also use Playwright Locator's [or](https://playwright.dev/docs/api/class-locator#locator-or) command so that you will end up `or`ing two locators for the Product description textbox, and these `or`ed locators will work on both "Visual" and "Text" mode.

I think either of those 2 approaches would work.

alvarothomas replied:
Nice one, thanks for suggesting to click Visual beforehand.

veljkho commented on plugins/woocommerce/tests/e2e-pw/tests/merchant/create-simple-product.spec.js:
```diff
@@ -57,7 +63,33 @@ test.describe.serial( 'Add New Simple Product Page', () => {
+
+		// Fill in purchase note
+		await page.getByText( 'Advanced' ).click();
+		await page.locator( '#_purchase_note' ).fill( productPurchaseNote );
```
I think you can use here `getByLabel( 'Purchase note' ).fill( productPurchaseNote );` and that would work nicely.

alvarothomas replied:
Good idea, thank you for suggesting that.

veljkho reviewed (commented):
I forgot to submit my review @alvarothomas, here are my comments down below.

rodelgc commented on plugins/woocommerce/tests/e2e-pw/tests/merchant/create-simple-product.spec.js:
```diff
@@ -57,7 +63,33 @@ test.describe.serial( 'Add New Simple Product Page', () => {
 		} );
 		await page.locator( '#title' ).fill( virtualProductName );
 		await page.locator( '#_regular_price' ).fill( productPrice );
+		await page.locator( '#_sale_price' ).fill( salePrice );
 		await page.locator( '#_virtual' ).click();
+
+		// Fill in a product description
+		await page
+			.frameLocator( '#content_ifr' )
+			.locator( '.wp-editor' )
+			.fill( productDescription );
```
It will also be good to use user-facing locators such as any of the `page.getBy*` ones, since this is one of the best practices that Playwright suggests [here](https://playwright.dev/docs/best-practices#prefer-user-facing-attributes-to-xpath-or-css-selectors).

As a tip, you can let Playwright pick the locator for you using [UI mode](https://playwright.dev/docs/test-ui-mode#opening-ui-mode), specifically its [locator picker](https://playwright.dev/docs/test-ui-mode#pick-locator). Although not entirely perfect, the locator picker defaults to giving you a user-facing locator you can use.

alvarothomas replied:
Thanks for this, @rodelgc !! It took me a while to understand how PW UI mode could help but, once I've got it, I've been able to use the suggestions to replace a bunch od locators with `getBy` instead. Please let me know what do you think about these changes when you can.

rodelgc commented on plugins/woocommerce/tests/e2e-pw/tests/merchant/create-simple-product.spec.js:
```diff
@@ -4,6 +4,12 @@ const wcApi = require( '@woocommerce/woocommerce-rest-api' ).default;
 const virtualProductName = 'Virtual Product Name';
 const nonVirtualProductName = 'Non Virtual Product Name';
 const productPrice = '9.99';
+const salePrice = '6.99';
+const productDescription = 'Virtual product description.';
+const productSKU = '1234567890'
+const productPurchaseNote = 'Virtual product purchase note'
+const productAttribute = 'color'
+const productAttributeColor = 'red | white'
```
I would suggest adding assertions for these values later on in the test.  So far, the only assertion we have on this test is that the "Product published" notification was displayed. Adding more assertions around these newly added details will make the test more, well... assertive 😅

alvarothomas replied:
Good idea @rodelgc , if I'm not mistaken, this aligns with my [comment - second point -](https://github.com/woocommerce/woocommerce-quality/issues/607#issuecomment-1898576008) in the related issue.

tl;dr: My impression is that this script automates the creation of a product, but there's no test afterwards to verify that the product was created with the expected values. I hope I'm explaining myself right, not 100% if this is what you meant here too.

Would you please mind sharing with me one example of assertion that can be used to verify any of those values?

rodelgc reviewed (dismissed):
Hi @alvarothomas I'm done reviewing the PR and requested some changes. 

This PR has a merge conflict, so kindly resolve that first before we merge it.

Thanks for working on this issue BTW! This gives the `can create simple virtual product` test more coverage. And great job learning our testing framework quickly!

Commit:
Update  add-virtual-product formatting and locators

Update  add-virtual-product tests with suggested formatting and locators

Commit:
Add changelog file

Adding the missing changelog file via pnpm --filter=plugin-woocommerce run changelog add

Commit:
Update locators with getBy on product creation

Update product name, price, sale price and virtual checkbox marking locators to use getBy instead

Commit:
Update SKU and publish locators with getBy instead

Updated SKU and publish locators with getBy instead

Commit:
Added asserts to create-virtual-prd test

Added asserts to create-virtual-prorduct test

Commit:
Update SKU check

Update SKU assertion

Commit:
Remove keyboard action

Remove press enter keyboard action

Commit:
Add changelog

Added changelog file

Commit:
Add changefile(s) from automation for the following project(s): woocommerce

Commit:
Update virtualProductName check

Update virtualProductName check to the const instead of the string

alvarothomas commented:
@rodelgc @veljkho I've updated the `can create simple virtual product` test with some asserts to verify the values in the product editor. If I'm not mistaken, the rest of the values are verified in the `can have a shopper add the simple non-virtual product to the cart` test too.

I'd need some assistance to resolve the conflict on the create-simple-product file. Would it be enough to just bring these in?

```
const productTag = 'nonVirtualTag';
const productCategory = 'nonVirtualCategory';
const productDescription = 'Description of a non-virtual product.';
const productDescriptionShort = 'Short description';
```

veljkho committed:
Rebase and format

veljkho reviewed (approved):
I rebased trunk and fixed conflict. @alvarothomas 
Worked greatly for me when I executed test. 👍
